core/src/main/java/apoc/trigger/Trigger.java

```diff
@@ -48,38 +54,61 @@ public TriggerInfo( String name, String query, Map<String,Object> selector, Map<
+    
+    @Context public Log log;
+
+    private void preprocessDeprecatedProcedures() {
+        final String msgDeprecation = "Please note that the current procedure is deprecated, \n" +
+                "it's recommended to use the `apoc.trigger.install`, `apoc.trigger.drop`, `apoc.trigger.dropAll`, `apoc.trigger.stop`, and `apoc.trigger.start` procedures \n" + 
+                "instead of, respectively, `apoc.trigger.add`, `apoc.trigger.remove`, `apoc.trigger.removeAll`, `apoc.trigger.pause`, and `apoc.trigger.resume`.";
+                
+        log.warn(msgDeprecation);
+        
+        if (!Util.isWriteableInstance(db, GraphDatabaseSettings.SYSTEM_DATABASE_NAME)) {
+            throw new RuntimeException(SYS_NON_LEADER_ERROR + msgDeprecation);
+        }
+    }
 
-    @Procedure(mode = Mode.WRITE)
+    @Deprecated
+    @Procedure(mode = Mode.WRITE, deprecatedBy = "apoc.trigger.install")
     @Description("add a trigger kernelTransaction under a name, in the kernelTransaction you can use {createdNodes}, {deletedNodes} etc., the selector is {phase:'before/after/rollback/afterAsync'} returns previous and new trigger information. Takes in an optional configuration.")
     public Stream<TriggerInfo> add(@Name("name") String name, @Name("kernelTransaction") String statement, @Name(value = "selector"/*, defaultValue = "{}"*/)  Map<String,Object> selector, @Name(value = "config", defaultValue = "{}") Map<String,Object> config) {
+        preprocessDeprecatedProcedures();
+        
         Util.validateQuery(db, statement);
         Map<String,Object> params = (Map)config.getOrDefault("params", Collections.emptyMap());
-        Map<String, Object> removed = triggerHandler.add(name, statement, selector, params);
-        if (removed != null) {
+        Map<String, Object> removed = TriggerUtils.add(db.databaseName(), name, statement, selector, params);
+        // always add transaction listener
+        triggerHandler.reconcileKernelRegistration(true);
+        if (!removed.isEmpty()) {
             return Stream.of(
                     new TriggerInfo(name,(String)removed.get("statement"), (Map<String, Object>) removed.get("selector"), (Map<String, Object>) removed.get("params"),false, false),
                     new TriggerInfo(name,statement,selector, params,true, false));
         }
         return Stream.of(new TriggerInfo(name,statement,selector, params,true, false));
     }
 
-    @Procedure(mode = Mode.WRITE)
+    @Deprecated
+    @Procedure(mode = Mode.WRITE, deprecatedBy = "apoc.trigger.drop")
     @Description("remove previously added trigger, returns trigger information")
     public Stream<TriggerInfo> remove(@Name("name")String name) {
-        Map<String, Object> removed = triggerHandler.remove(name);
-        if (removed == null) {
+        preprocessDeprecatedProcedures();
+        Map<String, Object> removed = TriggerUtils.remove(db.databaseName(), name);
+        triggerHandler.updateCache();
```

Comment:
Some bits have changed in the code here: for example the `updateCache` call. My personal preference would be to live all the old procedures as they were (even if we have to duplicate functionality temporarily). What do you think?

Reply:
Yes makes sense, I just split the trigger, now the deprecated one are in another class (`TriggerDeprecatedProcedures`  ) with their original code